### ISSUE TITLE
About dataset preprocessing

### ISSUE THREAD

**Issue by starhiking** [tony19980810/CausalNet#3]
Congratulations on your excellent work on CausalNet accepted at ICCV 2025! The proposed framework for robust MER against key-frame errors is very inspiring.

I am currently studying your code and attempting to reproduce the results, but I have encountered a few uncertainties regarding the dataset preparation and input definition. I would be very happy if you could provide some clarifications on the following points:

1. What's the difference between STSNET_whole_norm_u_v_os and three_norm_u_v_os folders?
2. What's the meaning of the PNG filename and subfolder names in the three_norm_u_v_os folders?
3. Do you plan to release the processing code for datasets to generate these optical flow images?
4. The paper emphasizes using the "representation of the entire ME sequence" to enhance robustness. Does this mean the model inputs a continuous sequence of optical flow frames (e.g., shape T x C x H x W), or does it rely on aggregated/sparse flows computed only between key-frames (e.g., one flow map for Onset-Apex and one for Apex-Offset)?

**Comment by tony19980810** [comment]
Hi starhiking,

Thank you for your interest in our work. Regarding your questions:

1. Here, STSNET_whole_norm_u_v_os is the complete sample set, while the three_norm_u_v_os folders are used for the LOSO (Leave-One-Subject-Out) training and testing splits. In this setup, we followed the training framework of HTNet.

2. The three_norm_u_v_os folders are indeed used for LOSO training and testing. During training, each subject's samples are individually and sequentially used as the test set, while the samples from all other subjects serve as the training set. This process is repeated for the number of subjects. Therefore, filenames such as '006', 'sub12', or 's01' represent subject IDs, while '0', '1', '2' denote labels. Taking the example of an image named 006_006_1_2: 006_006_1_2 represents the onset-to-apex optical flow. 006_006_1_2_1 represents the apex-to-offset optical flow. 006_006_1_2_2 represents the onset-to-apex optical flow direction. 006_006_1_2_3 represents the apex-to-offset optical flow direction.

3. Below is the core code we used to compute the optical flow. 

```
import cv2
import numpy as np
import pandas as pd

def pol2cart(rho, phi):
    x = rho * np.cos(phi)
    y = rho * np.sin(phi)
    return x, y

def computeStrain(u, v):
    u_x = u - pd.DataFrame(u).shift(-1, axis=1)
    v_y = v - pd.DataFrame(v).shift(-1, axis=0)
    u_y = u - pd.DataFrame(u).shift(-1, axis=0)
    v_x = v - pd.DataFrame(v).shift(-1, axis=1)
    os = np.array(np.sqrt(u_x**2 + v_y**2 + 0.25 * (u_y + v_x)**2).ffill(axis=1).ffill(axis=0))
    return os

def calc_os_flow(img1, img2):
    frame1 = cv2.imread(img1, 0)
    frame2 = cv2.imread(img2, 0)
    frame1= cv2.resize(frame1, (224, 224))
    frame2 = cv2.resize(frame2, (224, 224))
    optical_flow = cv2.optflow.DualTVL1OpticalFlow_create()
    flow = optical_flow.calc(frame1, frame2, None)
    magnitude, angle = cv2.cartToPolar(flow[..., 0], flow[..., 1])
    u, v = pol2cart(magnitude, angle)
    os_ = computeStrain(u, v)
    
    final_u = cv2.resize(u, (48, 48))
    final_v = cv2.resize(v, (48, 48))
    final_os = cv2.resize(os_, (48, 48))

    if ((np.max(final_u) - np.min(final_u))==0):
        normalized_u = final_u.astype(np.uint8)
    else:
        normalized_u = ((final_u - np.min(final_u)) / (np.max(final_u) - np.min(final_u)) * 255).astype(np.uint8)

    if ((np.max(final_v) - np.min(final_v))==0):
        normalized_v = final_v.astype(np.uint8)
    else:
        normalized_v = ((final_v - np.min(final_v)) / (np.max(final_v) - np.min(final_v)) * 255).astype(np.uint8)

    if ((np.max(final_os) - np.min(final_os))==0):
        normalized_os = final_os.astype(np.uint8)
    else:
        normalized_os = ((final_os - np.min(final_os)) / (np.max(final_os) - np.min(final_os)) * 255).astype(np.uint8)


    return np.concatenate(
            (normalized_os.reshape(*normalized_os.shape, 1), normalized_v.reshape(*normalized_v.shape, 1), normalized_u.reshape(*normalized_u.shape, 1)),
            axis=2
        )

of = calc_os_flow(onset_image_path,apex_image_path )
new_width, new_height = 28, 28
resized_of= cv2.resize(of, (new_width, new_height))
```


Since the preprocessing code involves steps such as face cropping and extracting optical flow based on keyframes, which is quite intricate, we do not plan to release the entire code for now. However, I recommend you check out this repository for reference:
https://github.com/GuanghaoZhu663/SKD-TSTSAN

4. We only input the optical flow from onset to apex and from apex to offset to represent the onset-to-offset micro-expression sequence.

I hope my answers are helpful to you. Once again, thank you for your attention to our work, and I wish you all the best in your research.

**Comment by starhiking** [comment]
Thanks！
